sagemathinc/cocalc
# Fix project upgrades

# Description
the overall goal is to limit the contributed upgrades for a project by the configured total amount of maximum upgrades. in particular, already applied upgrades by other users and the free quota were sort of ignored. there is also some consolidation happening, where duplicated code is collected in the same function.

this visualization helped me (it's in the source)

```
0                                total          maximum
|<-------------------------------->|                |
|<----->|<------------------------>|<-------------->|
| admin |  all upgrades by users   | proj remainder |
| +     |<------------>|<--------->|<--------->|    |
| free  |  other users | this user | remaining |    |
|       |              |           | this user |    |
|       |              |<--------------------->|    |
|       |              |  limit for this user  | <= | max
```

# Testing Steps
uff ... many special cases
1. checkbox-upgrades (binary 0/1): 
   * no upgrades → adding 1 is fine
   * user A has upgraded → switch to user B who also wants to upgrade → hitting limit
   * back to user A → remove the same upgrade  → should work
  1.1 like above, but instead of user B, do this for the admin (free quota)
1. non-negative number upgrades:
   * show how much available, and how much left to upgrade
   * also, we have to check the case where the total upgrade is already beyond the limit per project → this does immediately show an error. removing the upgrade removes that error (equivalent to clicking "max", because the max to apply is smaller than the current upgrade value by the user)
1. another small but important detail is to always check the enabled/disabled state of the save button. I don't remember exactly, but there was a situation where it didn't switch to being enabled although it should (probably because the limits are calculated correctly now). It should be enabled if there is indeed a change and none of the errors are active.
1. Here is a way to trigger "too many upgrades" errors: e.g. first (almost) max out mem or disk upgrades, then as admin, add even more upgrades beyond the max, and then edit as the user again. now, the user is told to reduce the upgrade to stay beyond the limit. e.g.

above limit:

![screenshot from 2018-11-22 16-41-02](https://user-images.githubusercontent.com/207405/48912253-7406ff00-ee75-11e8-9ed0-1b5a4f409daf.png)

error (save button disabled) -- the screenshot is older, error is different, see below:

![screenshot from 2018-11-22 16-41-10](https://user-images.githubusercontent.com/207405/48912261-7701ef80-ee75-11e8-933f-82a02b5bbac3.png)

and after clicking max, at the limit and save button enabled

![screenshot from 2018-11-22 16-41-18](https://user-images.githubusercontent.com/207405/48912265-79fce000-ee75-11e8-8076-d8dfcb41a707.png)

5. another interesting edge case is if the "admin" upgrades way beyond the upper limit. in that case, the user should be told to reduce upgrades to 0 (their limit), which violates the maximum of the project, but the save button should still be active (otherwise, the user wouldn't even be allowed to remove their quota).

6. also test improvements towards giving users feedback about why it is a problem. Below is a progressing sequence of error messages while being at the allowed limit (where some quotas are already applied -- similarly this should work if there are currently no quotas applied by the user), then exceeding the limit up until the amount of remaining quotas of the user, and then beyond the quotas of the user.

![screenshot from 2018-11-23 16-56-19](https://user-images.githubusercontent.com/207405/48951957-cd883000-ef40-11e8-8eb2-0a6acc335577.png)

![screenshot from 2018-11-23 16-56-26](https://user-images.githubusercontent.com/207405/48951959-d11bb700-ef40-11e8-923a-e9ab90431352.png)

![screenshot from 2018-11-23 16-56-38](https://user-images.githubusercontent.com/207405/48951964-d4af3e00-ef40-11e8-859c-9c43f8928372.png)

![screenshot from 2018-11-23 16-56-47](https://user-images.githubusercontent.com/207405/48951968-d7aa2e80-ef40-11e8-96c7-7249009b1bce.png)


6.1 similary, in the case where all user upgrades would bring the quota beyond the overall quota limit of the project, the messages report properly:

at the limit
![screenshot from 2018-11-23 16-26-43](https://user-images.githubusercontent.com/207405/48950827-c8c17d00-ef3c-11e8-9c86-4bea3048a269.png)

beyond
![screenshot from 2018-11-23 16-26-51](https://user-images.githubusercontent.com/207405/48950832-cced9a80-ef3c-11e8-9160-dec11e63f891.png)

at the maximum of how much quota the user has
![screenshot from 2018-11-23 16-27-11](https://user-images.githubusercontent.com/207405/48950838-cf4ff480-ef3c-11e8-88dd-d64e2b28cbe1.png)

and beyond that
![screenshot from 2018-11-23 16-27-18](https://user-images.githubusercontent.com/207405/48950841-d2e37b80-ef3c-11e8-8790-362696211132.png)


7. the component also appears in account settings → upgrades

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:
